tokio-boring/src/lib.rs

```diff
@@ -307,6 +307,58 @@ impl<S> HandshakeError<S> {
             _ => None,
         }
     }
+
+    /// Returns a reference to the inner SSL error, if this error was caused by
+    /// a SSL error.
+    pub fn as_ssl_error(&self) -> Option<&boring::error::ErrorStack> {
+        match &self.0 {
+            ssl::HandshakeError::Failure(s) => s.error().ssl_error(),
+            ssl::HandshakeError::SetupFailure(ref s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Consumes `self` and returns the inner I/O error by value, if this error
+    /// was caused by an I/O error.
+    ///
+    /// If this error was not caused by an I/O error, returns `Err(self)` so
+    /// that the error value can be reused.
+    pub fn into_io_error(self) -> Result<io::Error, Self> {
+        match self.0 {
+            ssl::HandshakeError::Failure(s) => {
+                if s.error().io_error().is_some() {
+                    Ok(s.into_error().into_io_error().expect(
+                        "if `s.error().io_error().is_some()`, `into_io_error()` must succeed",
+                    ))
```

Comment:
I think you could avoid the `expect()` here by adding a new `MidHandshakeSslStream::from_parts` function and using `into_parts()` above.